File: N/1019-next-greater-node-in-linked-list/main.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func nextLargerNodes(head *ListNode) []int {
	var stack []int
	var sidx []int
	var ret []int
	p := head
	for p != nil {
		ret = append(ret, 0)
		if len(stack) == 0 {
			stack = append(stack, p.Val)
			sidx = append(sidx, len(ret)-1)
		} else {
			idx := -1
			for j := len(stack) - 1; j >= 0; j-- {
				if stack[j] < p.Val {
					ret[sidx[j]] = p.Val
				} else {
					idx = j
					break
				}
			}

			stack = stack[0 : idx+1]
			sidx = sidx[0 : idx+1]

			stack = append(stack, p.Val)
			sidx = append(sidx, len(ret)-1)
		}
		p = p.Next
	}

	return ret
}

func main() {
	fmt.Println()
}

File: N/1019-next-greater-node-in-linked-list/main.rs
struct Solution {}

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
  pub val: i32,
  pub next: Option<Box<ListNode>>,
}

impl ListNode {
  #[inline]
  fn new(val: i32) -> Self {
    ListNode { next: None, val }
  }
}

impl Solution {
  pub fn next_larger_nodes(head: Option<Box<ListNode>>) -> Vec<i32> {
    let mut stack: Vec<i32> = Vec::new();
    let mut idx: Vec<usize> = Vec::new();
    let mut p = head;

    let mut ret: Vec<i32> = Vec::new();
    let mut offset: usize = 0;
    while let Some(sp) = p {
      while stack.len() > 0 && sp.val > stack[stack.len() - 1] {
        stack.pop();
        ret[idx.pop().unwrap()] = sp.val;
      }

      stack.push(sp.val);
      idx.push(offset);
      ret.push(0);
      offset += 1;
      p = sp.next;
    }

    ret
  }
}
